src/emulator/functionsRuntimeWorker.ts

```diff
@@ -47,24 +48,63 @@ export class RuntimeWorker {
             this.state = RuntimeWorkerState.IDLE;
           } else if (this.state === RuntimeWorkerState.FINISHING) {
             this.log(`IDLE --> FINISHING`);
-            this.runtime.shutdown();
+            this.runtime.process.kill();
           }
         }
       }
     });
 
-    this.runtime.exit.then(() => {
+    const childProc = this.runtime.process;
+    let msgBuffer = "";
+    childProc.on("message", (msg) => {
+      msgBuffer = this.processStream(msg, msgBuffer);
+    });
+
+    let stdBuffer = "";
+    if (childProc.stdout) {
+      childProc.stdout.on("data", (data) => {
+        stdBuffer = this.processStream(data, stdBuffer);
+      });
+    }
+
+    if (childProc.stderr) {
+      childProc.stderr.on("data", (data) => {
+        stdBuffer = this.processStream(data, stdBuffer);
+      });
+    }
+
+    childProc.on("exit", () => {
       this.log("exited");
       this.state = RuntimeWorkerState.FINISHED;
     });
   }
 
+  private processStream(s: Serializable, buf: string): string {
+    buf += s.toString();
+
+    const lines = buf.split("\n");
+    if (lines.length > 1) {
+      // slice(0, -1) returns all elements but the last
+      lines.slice(0, -1).forEach((line: string) => {
```

Comment:
I know this is just copied code, but wanted to call out that we prefer `for(...)` over `forEach(...)`

Reply:
Good new is that I throw away all this code in later PR! Will leave them unchanged in the PR.